src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+            Ping = 6,
+            GoAway = 7,
+            WindowUpdate = 8,
+            Continuation = 9
+        }
+
+        private struct FrameHeader
+        {
+            public readonly int Length;
+            public readonly FrameType Type;
+            public readonly FrameFlags Flags;
+            public readonly int StreamId;
+
+            public const int Size = 9;
+            public const int MaxLength = 16384;
+
+            public const int PriorityInfoLength = 5;       // for both PRIORITY frame and priority info within HEADERS
+            public const int PingLength = 8;
+            public const int WindowUpdateLength = 4;
+
+            public FrameHeader(int length, FrameType type, FrameFlags flags, int streamId)
+            {
+                Debug.Assert(length <= MaxLength);
+                Debug.Assert(streamId >= 0);
+
+                Length = length;
+                Type = type;
+                Flags = flags;
+                StreamId = streamId;
+            }
+
+            public bool PaddedFlag => (Flags & FrameFlags.Padded) != 0;
+            public bool AckFlag => (Flags & FrameFlags.Ack) != 0;
+            public bool EndHeadersFlag => (Flags & FrameFlags.EndHeaders) != 0;
+            public bool EndStreamFlag => (Flags & FrameFlags.EndStream) != 0;
+            public bool PriorityFlag => (Flags & FrameFlags.Priority) != 0;
+
+            public static FrameHeader Read(ReadOnlySpan<byte> buffer)
+            {
+                Debug.Assert(buffer.Length >= Size);
+
+                return new FrameHeader(
+                    (buffer[0] << 16) | (buffer[1] << 8) | buffer[2],
+                    (FrameType)buffer[3],
+                    (FrameFlags)buffer[4],
+                    (int)((uint)((buffer[5] << 24) | (buffer[6] << 16) | (buffer[7] << 8) | buffer[8]) & 0x7FFFFFFF));
+            }
+
+            public void Write(Span<byte> buffer)
+            {
+                Debug.Assert(buffer.Length >= Size);
+
+                buffer[0] = (byte)((Length & 0x00FF0000) >> 16);
+                buffer[1] = (byte)((Length & 0x0000FF00) >> 8);
+                buffer[2] = (byte)(Length & 0x000000FF);
+
+                buffer[3] = (byte)Type;
+                buffer[4] = (byte)Flags;
+
+                buffer[5] = (byte)((StreamId & 0xFF000000) >> 24);
+                buffer[6] = (byte)((StreamId & 0x00FF0000) >> 16);
+                buffer[7] = (byte)((StreamId & 0x0000FF00) >> 8);
+                buffer[8] = (byte)(StreamId & 0x000000FF);
+            }
+        }
+
+        [Flags]
+        private enum FrameFlags : byte
+        {
+            None = 0,
+            
+            // Some frame types define bits differently.  Define them all here for simplicity.
+
+            EndStream =     0b00000001,
+            Ack =           0b00000001,
+            EndHeaders =    0b00000100,
+            Padded =        0b00001000,
+            Priority =      0b00100000,
+        }
```

Comment:
Nit: consider separating all of these nested helper types out into their own files

Reply:
OK